File: pcd/processos/mpi/lista7/Exercicio2.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

#define tam 1.0
#define dx 0.00001
#define dt 0.000001
#define T    0.01
#define kappa 0.000045

void main(int argc, char* argv[]) {

    double *tmp, *u, *u_prev;
    double x, t, tempo, maxval, maxval_global;
    long int i, n, maxloc, ini, fim, maxloc_global;
    int rank, size, prox, ant;
    MPI_Status status;

    /* Inicializa o MPI */
    MPI_Init(&argc, &argv);

    tempo = MPI_Wtime();

    /* Detecta o rank do processo corrente e o total de processos */
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    /* Claculando quantidade de pontos */
    n = (tam/dx)/size;

    /* Alocando vetores */
    u = (double *) malloc((n + 1)*sizeof(double));
    u_prev = (double *) malloc((n + 1)*sizeof(double));

    /* Apenas o processo mestre imprime */
    if(rank == 0){
        /*printf("Inicio: qtde=%ld, dt=%g, dx=%g, dx²=%g, kappa=%f, const=%f\n",
                (n+2), dt, dx, dx*dx, kappa, kappa*dt/(dx*dx));
        printf("Iteracoes previstas: %g\n", T/dt);*/
    }

    /* Ajusta o valor inicial de x */
    if(rank != 0){
        x = rank * (n+1) * dx;
    }
    else{
        x = dx;
    }

    /* Cada um preenche seu pedaco de u_prev */
    for (i=1; i<n; i++) {
        if(x<=0.5){
            u_prev[i] = 200*x;
        }
        else{
            u_prev[i] = 200*(1.-x);
        }
        x += dx;
    }

    /* Calcula os processos com os quais interagir */
    prox = rank + 1;
    ant = rank - 1;

    /* Ajusta o proximo do ultimo e o anterior do primeiro */
    if(rank == 0){
        ant = MPI_PROC_NULL;
    }
    if(rank == size -1){
        prox = MPI_PROC_NULL;
    }

    t = 0.;
    while (t<T) {

        /* Calcula o novo u */
        x = dx;
        for (i=1; i<n; i++) {
            u[i] = u_prev[i] + kappa*dt/(dx*dx)*(u_prev[i-1]-2*u_prev[i]+u_prev[i+1]);
            if(u[i] >= 100) printf("%g\n", u[i]);
            x += dx;
        }

        /* Envia para o proximo e recebe do anterior */
        MPI_Sendrecv(&u[n-1], 1, MPI_DOUBLE, prox, 0,
                     &u_prev[0], 1, MPI_DOUBLE, ant, MPI_ANY_TAG,
                     MPI_COMM_WORLD, &status);

        /* Envia para o anterior e recebe do proximo */
        MPI_Sendrecv(&u[1], 1, MPI_DOUBLE, ant, 0,
                     &u_prev[n], 1, MPI_DOUBLE, prox, MPI_ANY_TAG,
                     MPI_COMM_WORLD, &status);

        /* Forca condicao de contorno */
        if(rank == 0){
            u[0] = 0.;
            u_prev[0] = 0.;
        }
        else if(rank == size-1){
            u[n] = 0.;
            u_prev[n] = 0.;
        }

        /* Copia tudo de u para u_prev */
        u[0] = u_prev[0];
        u[n] = u_prev[n];

        tmp = u_prev; u_prev = u; u = tmp; /* troca entre ponteiros */

        t += dt;
    }

    /* Calculando o maior valor e sua localizacao */
    maxloc = 1;
    for (i=1; i<n; i++) {
        if (u[i] > u[maxloc]){
            maxloc = i;
        }
    }

    maxval = u[maxloc];
    maxloc = (rank * (n+1)) + maxloc;

    /*printf("\nProcesso: %d\n\tmaxloc: %d\n\tmaxval: %g\n",
           rank, maxloc, maxval);*/

    if(rank != 0){
        MPI_Send(&maxval, 1, MPI_DOUBLE, 0, 0, MPI_COMM_WORLD);
        MPI_Send(&maxloc, 1, MPI_LONG, 0, 1, MPI_COMM_WORLD);
    }
    else{
        long int* indices = (long int*) malloc(size*sizeof(long int));
        double* valores = (double*) malloc(size*sizeof(double));

        valores[0] = maxval;
        indices[0] = maxloc;

        for(i=1; i<size; i++){
            MPI_Recv(&valores[0], 1, MPI_DOUBLE, i, 0, MPI_COMM_WORLD,
                     &status);
            MPI_Recv(&indices[i], 1, MPI_DOUBLE, i, 1, MPI_COMM_WORLD,
                     &status);
        }

        /*for(i=0; i<size; i++){
            printf("proc %d: u[%ld] = %lf\n", i, indices[i], valores[i]);
        }*/

        maxval_global = valores[0];
        maxloc_global = indices[0];
        for(i=1; i<size; i++){
            if(valores[i] > maxval_global){
                maxval_global = valores[i];
                maxloc_global = indices[i];
            }
        }

        tempo = MPI_Wtime() - tempo;
        printf("Tempo total: %.3lfs\n", tempo);
        printf("Maior valor u[%ld] = %g\n", maxloc_global, maxval_global);
    }

    MPI_Finalize();
}

File: pcd/processos/mpi/lista7/Exercicio2.f90
program ft2cs

include 'mpif.h'

! Const2ant2es
real, parameter :: TAM = 1.0
real, parameter :: DX = 0.00001
real, parameter :: DT = 0.000001
real, parameter :: T = 0.01
real, parameter :: KAPPA = 0.000045


! 'main'
real :: x, t2, tempo, maxval, maxval_global
integer :: i, n, maxloc, ini, fim, maxloc_global, rank, size, prox, ant2
integer :: status(MPI_STATUS_SIZE)

! inicia o MPI
call MPI_INIT(ierr)

call MPI_WTIME(tempo)

! Det2ect2a o rank do processo corrent2e e o total de processos
call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
call MPI_COMM_SIZE(MPI_COMM_WORLD, size, ierr)

! Calculando quant2idade de pont2os
n = (TAM/DX)/size

! Criando os vet2ores
real, dimension(n+1:1) :: u, u_prev, tmp

! Ajust2a o valor inicial de x
if(rank /= 0) then
	x = rank * (n+1) * DX
else
	x = DX
endif

! Cada um preenche seu pedaco de u_prev
do i=1,n,1
	if(x <= 0.5) then
		u_prev(i) = 200*x
	else
		u_prev(i) = 200*(1.-x)
	x += DX
enddo

! Calcula os processos com os quais int2eragir
prox = rank + 1
ant2 = rank - 1

! Ajust2a o proximo do ult2imo e o ant2erior do primeiro
if(rank == 0) then
	ant2 = MPI_PROC_NULL
else if(rank == size -1) then
	prox = MPI_PROC_NULL
endif

t2 = 0.0

while(t2 < T) do

	! Calcula o novo u
	x = dx
	do i=1, n, 1
		u(i) = u_prev(i) + KAPPA*DT/(DX*DX)*(u_prev(i-1)-2*u_prev(i)+u_prev(i+1))
		if(u(i) >= 100) then
			print2 *,u(i)
		endif
		x += DX
	enddo

	! Envia para o proximo e recebe do ant2erior
	call MPI_Sendrecv(&u(n-1), 1, MPI_DOUBLE, prox, 0, &
    	                 &u_prev(0), 1, MPI_DOUBLE, ant2, MPI_ANY_TAG, &
			             MPI_COMM_WORLD, status, ierr);

	! Envia para o ant2erior e recebe do proximo
    call MPI_Sendrecv(&u(1), 1, MPI_DOUBLE, ant2, 0, &
                     &u_prev(n), 1, MPI_DOUBLE, prox, MPI_ANY_TAG, &
                     MPI_COMM_WORLD, status, ierr);

	! Forca a condicao de cont2orno
	if(rank == 0) then
		u(0) = 0.0
		u_prev(0) = 0.0
	else if(rank == size-1) then
		u(n) = 0.0
		u_prev(n) = 0.0
	endif

	! Copia tudo de u para u_prev
	u(0) = u_prev(0)
	u(n) = u_prev(n)

	! Troca os vet2ores de lugar
	tmp = u_prev
	u_prev = u
	u = tmp

	t2 += DT
enddo

! Calculando o maior e sua localizacao
maxloc = 1
do i=1, n, 1
	if(u(i) > u(maxloc)) then
		maxloc = i
	endif
enddo

maxval = u(maxloc)
maxloc = (rank * (n+1) + maxloc)

if(rank /= 0) then
	call MPI_SEND(maxval, 1, MPI_DOUBLE, 0, 0, MPI_COMM_WORLD, ierr)
	call MPI_SEND(maxloc, 1, MPI_LONG, 0, 1, MPI_COMM_WOLD, ierr)
else
	integer indices(size)
	real valores(size)
endif

	valores(0) = maxval
	indices(0) = maxloc

	do i=1, size, 1
		call MPI_RECV(valores(i), 1, MPI_DOUBLE, i, 0, MPI_COMM_WORLD, status, ierr)
		call MPI_RECV(indices(i), 1, MPI_DOUBLE, i, 1, MPI_COMM_WORLD, status, ierr)
	enddo

	maxval_global = valores(0)
	maxloc_global = indices(0)

	do i=1, size, 1
		if(valores(i) > maxval_global) then
			maxval_global = valores(i)
			maxloc_global = indices(i)
		endif
	enddo

	real :: aux
	call MPI_WTIME(aux)
	tempo = aux - tempo

	print2 *,'Tempo total: ', tempo, '\n'
	print2 *,'Maior valor u(', maxloc_global, ') = ', maxval_global 

	call MPI_FINALIZE(ierr)

nd program ft2cs
